instrumentation/http/src/main/java/brave/http/HttpParserAdapter.java

```diff
@@ -0,0 +1,111 @@
+@Deprecated final class HttpParserAdapter implements HttpRequestParser, HttpResponseParser {
+  final CurrentTraceContext currentTraceContext;
+  final HttpParser parser;
+
+  HttpParserAdapter(CurrentTraceContext currentTraceContext, HttpParser parser) {
+    this.currentTraceContext = currentTraceContext;
+    this.parser = parser;
+  }
+
+  @Override public void parse(HttpRequest request, TraceContext context, SpanCustomizer span) {
+    HttpAdapter<Object, ?> adapter;
+    Object req;
+
+    // The order matters here because the deprecated HttpClientHandler.create(httpTracing, adapter)
+    // can be used even with the new HttpRequest type. This means we have to check the HttpResponse
+    // type after checking to see if we are in an adapter scenario.
+    if (request instanceof HttpClientRequest.FromHttpAdapter) {
+      HttpClientRequest.FromHttpAdapter wrapped = (HttpClientRequest.FromHttpAdapter) request;
+      adapter = wrapped.adapter;
+      req = wrapped.request;
+    } else if (request instanceof HttpServerRequest.FromHttpAdapter) {
+      HttpServerRequest.FromHttpAdapter wrapped = (HttpServerRequest.FromHttpAdapter) request;
+      adapter = wrapped.adapter;
+      req = wrapped.request;
+    }
+    if (request instanceof HttpClientRequest) {
+      req = request.unwrap();
+      if (req == null) return; // unwrap() shouldn't return null, but we shouldn't crash either.
+      adapter = new HttpClientRequest.ToHttpAdapter((HttpClientRequest) request);
+    } else if (request instanceof HttpServerRequest) {
+      req = request.unwrap();
+      if (req == null) return; // unwrap() shouldn't return null, but we shouldn't crash either.
+      adapter = new HttpServerRequest.ToHttpAdapter((HttpServerRequest) request);
+    } else {
+      throw new AssertionError("programming bug");
+    }
+
+    Scope ws = currentTraceContext.maybeScope(context);
+    try {
+      parser.request(adapter, req, span);
+    } finally {
+      ws.close();
+    }
+  }
+
+  @Override public void parse(@Nullable HttpResponse response, @Nullable Throwable error,
+    TraceContext context, SpanCustomizer span) {
+    HttpAdapter<?, Object> adapter;
+    Object res;
+
+    // The order matters here because the deprecated HttpClientHandler.create(httpTracing, adapter)
+    // can be used even with the new HttpResponse type. This means we have to check the HttpResponse
+    // type after checking to see if we are in an adapter scenario.
+    if (response instanceof HttpClientResponse.FromHttpAdapter) {
+      HttpClientResponse.FromHttpAdapter wrapped = (HttpClientResponse.FromHttpAdapter) response;
+      adapter = wrapped.adapter;
+      res = wrapped.response;
+    } else if (response instanceof HttpServerResponse.FromHttpAdapter) {
+      HttpServerResponse.FromHttpAdapter wrapped = (HttpServerResponse.FromHttpAdapter) response;
+      adapter = wrapped.adapter;
+      res = wrapped.response;
+    } else if (response instanceof HttpClientResponse) {
```

Comment:
Is it ok to have the `else`? (we don't have it above)

Reply:
because HttpClientAdapters.FromResponseAdapter is an instanceof HttpClientResponse I think we wouldn't intentionally break the if/else chain. as that would cause us to re-enter. I put stronger comments above this which can be improved if you feel need be.